### DISCUSSION TITLE
Endpoints are not looking clean. Is it me or just serverpod?

### DISCUSSION THREAD

**Discussion by qiqetes** [serverpod/serverpod#2570] [Q&A] [answered]
I have a problem that I haven't got with other frameworks, to check if a user can use a certain route I have to write a lot of code that I repeat on lots of endpoints. I want to see if there are better ways to do it. Thanks!
```dart
 Future<List<ClimbSend?>> getGradesForCountry(
      Session session, Conuntry country, int userId) async {
    final UserCheck userCheck = UserCheck(session);
    userCheck.isLoggedIn();

    final int requestingUserId = await userCheck.isAuthenticated();

    // 1. User in Contry?
    final userCountry = await UserCountry.db.findFirstRow(session, where: (t) {
      return t.userId.equals(requestingUserId) & t.country.equals(country.id);
    });

    if (userCountry == null) {
      throw Exception("User does not pertain to this country");
    }
    // ... more
}```

**Comment by vlidholt** [comment, marked as answer]
Yes, there is a much better way to do this. Serverpod has built in support for handling authentication on a per Endpoint basis. Just override the `requireLogin` property of the endpoint. Then access will be denied to any of the methods if the user isn't signed in. You also have the options to add scopes to limit access to a specific type of user (e.g., an admin).

```dart
class MyEndpoint extends Endpoint {
  @override
  bool get requireLogin => true;

  Future<void> myMethod(Session session) async {
    ...
  }
  ...
}
```

Read more here: https://docs.serverpod.dev/concepts/authentication/basics

**Reply by qiqetes** [reply]
But I don't only need to get if it's authenticated. What I need:
1. Is the user authenticated? Ok
2. Is the user in Group? (there are plenty of groups, so it can't be a scope)

For the userId (for later use to check if it's allowed in a certain group) I need to check if it's authenticated and that's an async operation that will worsen the performance.

```dart
 Future<int> isAuthenticated() async {
    final authenticated = await session.authenticated;
    final int? userId = authenticated?.userId;
    if (userId == null) {
      throw Exception("User logged not found");
    }

    return userId;
  }
  ```
  I'm using this, but again, I hope there's a better option
